libsolutil/LP.cpp

```diff
@@ -0,0 +1,806 @@
+			else
+				row = i;
+		}
+		else if (_tableau.data[i][_column] != 0)
+			return std::nullopt;
+	return row;
+}
+
+/// @returns a solution vector, assuming one exists.
+/// The solution vector minimizes the objective function if the tableau
+/// is the result of the simplex algorithm.
+vector<rational> solutionVector(Tableau const& _tableau)
+{
+	vector<rational> result;
+	vector<bool> rowsSeen(_tableau.data.size(), false);
+	for (size_t j = 1; j < _tableau.objective.size(); j++)
+	{
+		optional<size_t> row = basisIndex(_tableau, j);
+		if (row && rowsSeen[*row])
+			row = nullopt;
+		result.emplace_back(row ? _tableau.data[*row][0] : rational{});
+		if (row)
+			rowsSeen[*row] = true;
+	}
+	return result;
+}
+
+
+/// Solve the LP A x = b s.t. min c^T x
+/// Here, c is _tableau.objective and the first column of _tableau.data
+/// encodes b and the other columns encode A
+/// Assumes the tableau has a trivial basic feasible solution.
+/// Tries for a number of iterations and then gives up.
+pair<LPResult, Tableau> simplexEq(Tableau _tableau)
+{
+	size_t const iterations = min<size_t>(60, 50 + _tableau.objective.size() * 2);
+	for (size_t step = 0; step <= iterations; ++step)
+	{
+		optional<size_t> pivotColumn = findPivotColumn(_tableau);
+		if (!pivotColumn)
+			return make_pair(LPResult::Feasible, move(_tableau));
+		optional<size_t> pivotRow = findPivotRow(_tableau, *pivotColumn);
+		if (!pivotRow)
+			return make_pair(LPResult::Unbounded, move(_tableau));
+		performPivot(_tableau, *pivotRow, *pivotColumn);
+	}
+	return make_pair(LPResult::Unknown, Tableau{});
+}
+
+/// We add slack variables to find a basic feasible solution.
+/// In particular, there is a slack variable for each row
+/// which is weighted negatively. Setting the new slack
+/// variables to one and all other variables to zero yields
+/// a basic feasible solution.
+/// If the optimal solution has all slack variables set to zero,
+/// this is a basic feasible solution. Otherwise, the original
+/// problem is infeasible.
+/// This function returns the modified tableau with the original
+/// objective function and the slack variables removed.
+pair<LPResult, Tableau> simplexPhaseI(Tableau _tableau)
+{
+	LinearExpression originalObjective = _tableau.objective;
+
+	size_t rows = _tableau.data.size();
+	size_t columns = _tableau.objective.size();
+	for (size_t i = 0; i < rows; ++i)
+	{
+		if (_tableau.data[i][0] < 0)
+			_tableau.data[i] *= -1;
+		_tableau.data[i].resize(columns + rows);
+		_tableau.data[i][columns + i] = 1;
+	}
+	_tableau.objective = {};
+	_tableau.objective.resize(columns);
+	_tableau.objective.resize(columns + rows, rational{-1});
+
+	// This sets the objective factors of the slack variables
+	// to zero (and thus selects a basic feasible solution).
+	selectLastVectorsAsBasis(_tableau);
+
+	LPResult result;
+	tie(result, _tableau) = simplexEq(move(_tableau));
+	solAssert(result == LPResult::Feasible || result == LPResult::Unbounded, "");
```

Comment:
This constraint input: https://gist.github.com/bshastry/f37681fea0447356d1dcf000d18413ed

throws here

Reply:
Minimised input: https://gist.github.com/bshastry/b3a67e44b40b69aaaa1540ab0c96c7ef

where each line is to be parsed as `constraintType, rhsConstant, coefficients...`

Reply:
oh wow! Of course the result can also be `Unknown` - simply when it takes too long to find the solution...